### ISSUE TITLE
[nodetool]: Scylla ignores the parallelism configuration 

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#14691]
*Installation details*
HEAD: 780aee956857eb527fb20bbc1cc6901741dfaab0

*Description*
Scylla completely ignores the `-seq` configurable of the `nodetool repair` and will still run a repair with the maximum parallelism:

```
$ nodetool repair -pr ks cf
[2023-07-13 12:25:54,960] Starting repair command #1, repairing 1 ranges for keyspace ks (parallelism=SEQUENTIAL, full=true)
```

And in the syslog we see:

```
Jul 13 12:25:54 hostname-1 scylla[1911908]:  [shard  0] repair - repair[24564ac1-e25f-4d01-abd6-3d24f40f9079]: starting user-requested repair for keyspace ks, repair id 1, options {{ trace -> false}, { columnFamilies -> cf}, { jobThreads -> 1}, { incremental -> false}, { parallelism -> parallel}, { primaryRange -> true}}
Jul 13 12:25:54 hostname-1 scylla[1911908]:  [shard  0] repair - primary-range repair
```

Note the `parallelism -> parallel`

And scylla will effectively repair the `max_repair_ranges_in_parallel` ranges.

Being able to control the repair parallelism is an absolutely critical ability, especially when repair misbehaves.

**Comment by vladzcloudius** [comment]
cc @asias @mykaul @bhalevy

**Comment by asias** [comment]
The option -seq is not what you think it is. The -seq option does not make sense to Scylla since we do not use Merkle tree based repair since day one. 

https://dba.stackexchange.com/questions/156739/cassandra-nodetool-repair-parallelism-parallel-versus-dc-parallel

That's why it is ignored. 

If we want to repair less ranges in parallel, specify less ranges to repair per request. For example, specify one range at a time, this way, scylla will only repair one range in parallel.

**Comment by bhalevy** [comment]
Indeed it's not documented in https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html#:~:text=Repair%20%2D%20a%20process%20that%20runs,its%20replications%20will%20be%20repaired.

But rather than ignoring it we should return an error (probably from scylla-jmx)

**Comment by tchaikov** [comment]
we might want to print a warning in https://github.com/scylladb/scylla-tools-java/blob/9f63a96f11b1309a11b63d030ed9ac47d942df9b/src/java/org/apache/cassandra/tools/nodetool/Repair.java#L107-L108 

and *can* also throw an exception in https://github.com/scylladb/scylladb/blob/dad5caf1417362ed136e19db9f21dc6b3ebdbd90/repair/repair.cc#L808-L815 . but that'd break existing tooling if users are using this option.

**Comment by DoronArazii** [comment]
@eliransin please update the documantation

**Comment by bhalevy** [comment]
> @eliransin please update the documantation

[Documentation](https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html#:~:text=Repair%20%2D%20a%20process%20that%20runs,its%20replications%20will%20be%20repaired.) doesn't mention neither the `-seq` option nor the `-j jobs` option, but those appear in our API doc:
https://github.com/scylladb/scylladb/blob/a93fd2b162a80db905519d14e10f7b621bdd8dc1/api/api-doc/storage_service.json#L1030-L1037
https://github.com/scylladb/scylladb/blob/a93fd2b162a80db905519d14e10f7b621bdd8dc1/api/api-doc/storage_service.json#L1046-L1053

And they are parsed in https://github.com/scylladb/scylladb/blob/a93fd2b162a80db905519d14e10f7b621bdd8dc1/api/storage_service.cc#L321-L323

But `repair_service::do_repair_start` ignores some options rather than returning an error for unsupported options in `options_map`.

IMO we should simply return an error from `repair_service::do_repair_start` if any unsupported option is set (in a way we don't support, e.g. if `-j` is set by any uppoer layer to 1 by default, it should be supported of course).

**Comment by asias** [comment]
> > The option -seq is not what you think it is. The -seq option does not make sense to Scylla since we do not use Merkle tree based repair since day one.
> > https://dba.stackexchange.com/questions/156739/cassandra-nodetool-repair-parallelism-parallel-versus-dc-parallel
> > That's why it is ignored.
> > If we want to repair less ranges in parallel, specify less ranges to repair per request. For example, specify one range at a time, this way, scylla will only repair one range in parallel.
> 
> Thanks, @asias. I indeed misunderstood the semantics of the `-sec` parameter. And I think that the parameter that I had in mind is `-j job_threads`.
> 
> And I see we have both in our REST APIs. For example `/storage_service/repair_async/{keyspace}` has:
> 
> ```
> {
>                      "name":"parallelism",
>                      "description":"Repair parallelism, can be 0 (sequential), 1 (parallel) or 2 (datacenter-aware).",
>                      "required":false,
>                      "allowMultiple":false,
>                      "type":"string",
>                      "paramType":"query"
>                   },
> ...
> {
>                      "name":"jobThreads",
>                      "description":"An integer specifying the parallelism on each node.",
>                      "required":false,
>                      "allowMultiple":false,
>                      "type":"string",
>                      "paramType":"query"
>                   },
> ```
> 
> And the later seems to be the one that is supposed to control the internal parallelism of a specific repair job, isn't it?
> 
> Do we respect its value or ignore it together with `parallelism`?

Both parallelism and jobThreads are ignored. They are present for compatibility reasons, e.g., reqeust from nodetool / jmx won't get rejected. 

I had a PR https://github.com/scylladb/scylladb/pull/4848 (repair: Add ranges_parallelism option #4848) to allow specify the parallelism explicitly for advanced users, but it was rejected by maintainers.

**Comment by asias** [comment]
> > @eliransin please update the documantation
> 
> [Documentation](https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html#:~:text=Repair%20%2D%20a%20process%20that%20runs,its%20replications%20will%20be%20repaired.) doesn't mention neither the `-seq` option nor the `-j jobs` option, but those appear in our API doc:
> 
> https://github.com/scylladb/scylladb/blob/a93fd2b162a80db905519d14e10f7b621bdd8dc1/api/api-doc/storage_service.json#L1030-L1037
> 
> https://github.com/scylladb/scylladb/blob/a93fd2b162a80db905519d14e10f7b621bdd8dc1/api/api-doc/storage_service.json#L1046-L1053
> 
> And they are parsed in
> 
> https://github.com/scylladb/scylladb/blob/a93fd2b162a80db905519d14e10f7b621bdd8dc1/api/storage_service.cc#L321-L323
> 
> But `repair_service::do_repair_start` ignores some options rather than returning an error for unsupported options in `options_map`.
> 
> IMO we should simply return an error from `repair_service::do_repair_start` if any unsupported option is set (in a way we don't support, e.g. if `-j` is set by any uppoer layer to 1 by default, it should be supported of course).

The problem with being strict about the options is that old tools will break in a very unpleasant way.

**Comment by bhalevy** [comment]
> The problem with being strict about the options is that old tools will break in a very unpleasant way.

Why is this a problem?  We can't infinite backward compatibility to drag us backwards.
We provide a version of nodetool of our own.
If someone needs to, they can use our own version, that shouldn't break.
FE / R&D or a naive user wasting time on trying to use an unsupported tool is worse.
It's not even documented that we don't support those...
How would anyone know?  Ask you? Read the code?

**Comment by tchaikov** [comment]
i still think the way to address this is to print a warning in the java tool, it should minimize the suprise. if we can assume that user always use the vendored nodetool, then to remove the option in .json and throw an exception in the option parser is a more correct way to handle this option.

**Comment by asias** [comment]
> > The problem with being strict about the options is that old tools will break in a very unpleasant way.
> 
> Why is this a problem? We can't infinite backward compatibility to drag us backwards. We provide a version of nodetool of our own. If someone needs to, they can use our own version, that shouldn't break. FE / R&D or a naive user wasting time on trying to use an unsupported tool is worse. It's not even documented that we don't support those... How would anyone know? Ask you? Read the code?

Imagine a user upgrades scylla server but not nodetool. All nodetool repair fails. Why do you want to break this compatibility?

None of our document claims the -seq is supported: https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html.

We can start to modify nodetool to get rid of those unsupported options. We can also mark those api parameters deprecated. But I do not think starting by rejecting unsupported existing API parameter is a good idea.

**Comment by bhalevy** [comment]
> i still think the way to address this is to print a warning in the java tool, it should minimize the surprise while keep the compatibility of existing tooling of user. if we can assume that user always use the vendored nodetool, then to remove the option in .json and throw an exception in the option parser is a more correct way to handle this option.

We eventually want to move away from the java nodetool so adding a warning sounds reasonable.

I didn't suggest to remove those options from the .json definition, but rather return errors from `repair_service::do_repair_start` (or from `repair_options::repair_options`) when they contain values we don't support.
e.g. extending https://github.com/scylladb/scylladb/blob/567b4536892f5e2f842189b4d022941070328150/repair/repair.cc#L813-L815

I think that nodetool passes `--jobs 1` so we don't want to delete the option entirely, but we can return an error, or at least print a big fat warning if number of jobs is different than one.

> > > The problem with being strict about the options is that old tools will break in a very unpleasant way.
> > 
> > 
> > Why is this a problem? We can't infinite backward compatibility to drag us backwards. We provide a version of nodetool of our own. If someone needs to, they can use our own version, that shouldn't break. FE / R&D or a naive user wasting time on trying to use an unsupported tool is worse. It's not even documented that we don't support those... How would anyone know? Ask you? Read the code?
> 
> Imagine a user upgrades scylla server but not nodetool. All nodetool repair fails. Why do you want to break this compatibility?

I don't.  I want it to fail only if it tries to do something we don't currently implement.

> 
> None of our document claims the -seq is supported: https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html.
> 
> We can start to modify nodetool to get rid of those unsupported options. We can also mark those api parameters deprecated. But I do not think starting by rejecting unsupported existing API parameter is a good idea.

I think that first mediation is to add to the documentation a paragraph about legacy options we do not support,
and a paragraph about how to control repair parallelism with the available options.

About rejecting options, please see above.
The point is not rejecting them as a syntax error, but rather rejecting values we don't support on the semantic level.

**Comment by vladzcloudius** [comment]
This GH issue took a very unexpected turn.
Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla?
Like in https://github.com/scylladb/scylladb/pull/4848.
Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?

How hard that would be?

**Comment by asias** [comment]
> This GH issue took a very unexpected turn. Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla? 

Reusing the nodetool options in a different meaning is even more confusing for users with experience stock nodetool. For example:

```
        -j <job_threads>, --job-threads <job_threads>
            Number of threads to run repair jobs. Usually this means number of
            CFs to repair concurrently. WARNING: increasing this puts more load
            on repairing nodes, so be careful. (default: 1, max: 4)
```

This is completely different from what you wanted the -j option to be below. 

Instead of playing this game, IMO, we should deprecate nodetool all together and replace it with a more modern and clean tool.  

> Like in #4848. Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?
> 
> How hard that would be?

Note nodetool is a single node tool. It is not a cluster wide management tool to schedule repair plans. Extending nodetool to do cluster wide management is way beyond the scope of this GH issue.

**Comment by vladzcloudius** [comment]
> > This GH issue took a very unexpected turn. Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla?
> 
> Reusing the nodetool options in a different meaning is even more confusing for users with experience stock nodetool. For example:
> 
> ```
>         -j <job_threads>, --job-threads <job_threads>
>             Number of threads to run repair jobs. Usually this means number of
>             CFs to repair concurrently. WARNING: increasing this puts more load
>             on repairing nodes, so be careful. (default: 1, max: 4)
> ```
> 
> This is completely different from what you wanted the -j option to be below.
> 
> Instead of playing this game, IMO, we should deprecate nodetool all together and replace it with a more modern and clean tool.
> 
> > Like in #4848. Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?
> > How hard that would be?
> 
> Note nodetool is a single node tool. It is not a cluster wide management tool to schedule repair plans. Extending nodetool to do cluster wide management is way beyond the scope of this GH issue.

We already have such a tool - called Scylla Manager.
However, as you very well know, sometimes you still need an alternative.

And it can be either a `nodetool` or a naked REST API. However we need a simple way to control the max repair concurrency (among other things).
Why not to implement that? We know very well how to do that.

**Comment by asias** [comment]
> > > This GH issue took a very unexpected turn. Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla?
> > 
> > 
> > Reusing the nodetool options in a different meaning is even more confusing for users with experience stock nodetool. For example:
> > ```
> >         -j <job_threads>, --job-threads <job_threads>
> >             Number of threads to run repair jobs. Usually this means number of
> >             CFs to repair concurrently. WARNING: increasing this puts more load
> >             on repairing nodes, so be careful. (default: 1, max: 4)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is completely different from what you wanted the -j option to be below.
> > Instead of playing this game, IMO, we should deprecate nodetool all together and replace it with a more modern and clean tool.
> > > Like in #4848. Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?
> > > How hard that would be?
> > 
> > 
> > Note nodetool is a single node tool. It is not a cluster wide management tool to schedule repair plans. Extending nodetool to do cluster wide management is way beyond the scope of this GH issue.
> 
> We already have such a tool - called Scylla Manager. However, as you very well know, sometimes you still need an alternative.
> 
> And it can be either a `nodetool` or a naked REST API. However we need a simple way to control the max repair concurrency (among other things). Why not to implement that? We know very well how to do that.

We can already do that with the existing restful api, i.e., issue a request with a single range. The value of allows specifying more ranges but allow only one range to work in parallel interally by introducing a new api parameter is limited, because 1) for advanced users, one range per request achieves the goal already (using the old repair script or write a new one) 2) normal user will not use this option at all.  

A more advanced parameter like in #4848, which bypasses the system max_repair_ranges_in_parallel, allowing whatever parallelism specified by user. On one hand, it is more flexible for advanced users. On the other hand, it could cause problems when misused.

**Comment by vladzcloudius** [comment]
> > > > This GH issue took a very unexpected turn. Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla?
> > > 
> > > 
> > > Reusing the nodetool options in a different meaning is even more confusing for users with experience stock nodetool. For example:
> > > ```
> > >         -j <job_threads>, --job-threads <job_threads>
> > >             Number of threads to run repair jobs. Usually this means number of
> > >             CFs to repair concurrently. WARNING: increasing this puts more load
> > >             on repairing nodes, so be careful. (default: 1, max: 4)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This is completely different from what you wanted the -j option to be below.
> > > Instead of playing this game, IMO, we should deprecate nodetool all together and replace it with a more modern and clean tool.
> > > > Like in #4848. Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?
> > > > How hard that would be?
> > > 
> > > 
> > > Note nodetool is a single node tool. It is not a cluster wide management tool to schedule repair plans. Extending nodetool to do cluster wide management is way beyond the scope of this GH issue.
> > 
> > 
> > We already have such a tool - called Scylla Manager. However, as you very well know, sometimes you still need an alternative.
> > And it can be either a `nodetool` or a naked REST API. However we need a simple way to control the max repair concurrency (among other things). Why not to implement that? We know very well how to do that.
> 
> We can already do that with the existing restful api, i.e., issue a request with a single range. The value of allows specifying more ranges but allow only one range to work in parallel interally by introducing a new api parameter is limited, because 1) for advanced users, one range per request achieves the goal already (using the old repair script or write a new one) 2) normal user will not use this option at all.
> 
> A more advanced parameter like in #4848, which bypasses the system max_repair_ranges_in_parallel, allowing whatever parallelism specified by user. On one hand, it is more flexible for advanced users. On the other hand, it could cause problems when misused.

@asias that's the whole point - this GH issue is not targeting "advanced users". It targets "simple people", like myself, who want to issue a single repair command and not bother who scylla does it's magic.
And the same with restricting the parallelism: use a trivial "human readable" parameter to tell how many repairs per shard are allowed and let scylla handle the rest.

Why can't we do that? (Again) Seems to be simple enough.

**Comment by asias** [comment]
> > > > > This GH issue took a very unexpected turn. Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla?
> > > > 
> > > > 
> > > > Reusing the nodetool options in a different meaning is even more confusing for users with experience stock nodetool. For example:
> > > > ```
> > > >         -j <job_threads>, --job-threads <job_threads>
> > > >             Number of threads to run repair jobs. Usually this means number of
> > > >             CFs to repair concurrently. WARNING: increasing this puts more load
> > > >             on repairing nodes, so be careful. (default: 1, max: 4)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > This is completely different from what you wanted the -j option to be below.
> > > > Instead of playing this game, IMO, we should deprecate nodetool all together and replace it with a more modern and clean tool.
> > > > > Like in #4848. Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?
> > > > > How hard that would be?
> > > > 
> > > > 
> > > > Note nodetool is a single node tool. It is not a cluster wide management tool to schedule repair plans. Extending nodetool to do cluster wide management is way beyond the scope of this GH issue.
> > > 
> > > 
> > > We already have such a tool - called Scylla Manager. However, as you very well know, sometimes you still need an alternative.
> > > And it can be either a `nodetool` or a naked REST API. However we need a simple way to control the max repair concurrency (among other things). Why not to implement that? We know very well how to do that.
> > 
> > 
> > We can already do that with the existing restful api, i.e., issue a request with a single range. The value of allows specifying more ranges but allow only one range to work in parallel interally by introducing a new api parameter is limited, because 1) for advanced users, one range per request achieves the goal already (using the old repair script or write a new one) 2) normal user will not use this option at all.
> > A more advanced parameter like in #4848, which bypasses the system max_repair_ranges_in_parallel, allowing whatever parallelism specified by user. On one hand, it is more flexible for advanced users. On the other hand, it could cause problems when misused.
> 
> @asias that's the whole point - this GH issue is not targeting "advanced users". It targets "simple people", like myself, who want to issue a single repair command and not bother who scylla does it's magic. And the same with restricting the parallelism: use a trivial "human readable" parameter to tell how many repairs per shard are allowed and let scylla handle the rest.
> 
> Why can't we do that? (Again) Seems to be simple enough.

Normal users would not care about the internal parallelism has to be 1. The automatic parallelism should just work out of the box.  If it is not, we should fix the issues at the scylla core, e.g., the deadlock between the view update and streaming reader semaphore. So this is actually targeting advanced user cases to use this to workaround something. And you are not considered as "simple people" from any perspective ;-)

**Comment by vladzcloudius** [comment]
> > > > > > This GH issue took a very unexpected turn. Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla?
> > > > > 
> > > > > 
> > > > > Reusing the nodetool options in a different meaning is even more confusing for users with experience stock nodetool. For example:
> > > > > ```
> > > > >         -j <job_threads>, --job-threads <job_threads>
> > > > >             Number of threads to run repair jobs. Usually this means number of
> > > > >             CFs to repair concurrently. WARNING: increasing this puts more load
> > > > >             on repairing nodes, so be careful. (default: 1, max: 4)
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > This is completely different from what you wanted the -j option to be below.
> > > > > Instead of playing this game, IMO, we should deprecate nodetool all together and replace it with a more modern and clean tool.
> > > > > > Like in #4848. Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?
> > > > > > How hard that would be?
> > > > > 
> > > > > 
> > > > > Note nodetool is a single node tool. It is not a cluster wide management tool to schedule repair plans. Extending nodetool to do cluster wide management is way beyond the scope of this GH issue.
> > > > 
> > > > 
> > > > We already have such a tool - called Scylla Manager. However, as you very well know, sometimes you still need an alternative.
> > > > And it can be either a `nodetool` or a naked REST API. However we need a simple way to control the max repair concurrency (among other things). Why not to implement that? We know very well how to do that.
> > > 
> > > 
> > > We can already do that with the existing restful api, i.e., issue a request with a single range. The value of allows specifying more ranges but allow only one range to work in parallel interally by introducing a new api parameter is limited, because 1) for advanced users, one range per request achieves the goal already (using the old repair script or write a new one) 2) normal user will not use this option at all.
> > > A more advanced parameter like in #4848, which bypasses the system max_repair_ranges_in_parallel, allowing whatever parallelism specified by user. On one hand, it is more flexible for advanced users. On the other hand, it could cause problems when misused.
> > 
> > 
> > @asias that's the whole point - this GH issue is not targeting "advanced users". It targets "simple people", like myself, who want to issue a single repair command and not bother who scylla does it's magic. And the same with restricting the parallelism: use a trivial "human readable" parameter to tell how many repairs per shard are allowed and let scylla handle the rest.
> > Why can't we do that? (Again) Seems to be simple enough.
> 
> Normal users would not care about the internal parallelism has to be 1. The automatic parallelism should just work out of the box. 

Exactly. But if it was 1 by default I would probably not even open this issue since it would "just work". But it seems it's not 1 but rather `max_repair_ranges_in_parallel` and this is exactly why I started this discussion.

"just work" or a way to control it otherwise is exactly what I'm asking here for.

> If it is not, we should fix the issues at the scylla core, e.g., the deadlock between the view update and streaming reader semaphore. So this is actually targeting advanced user cases to use this to workaround something. And you are not considered as "simple people" from any perspective ;-)

I'm speaking for "simple people" here... ;)

**Comment by asias** [comment]
> > > > > > > This GH issue took a very unexpected turn. Let's try to get it back to the issue at hand: why don't we DO support these arguments (as in "stop ignoring them") giving them the semantics that suits Scylla?
> > > > > > 
> > > > > > 
> > > > > > Reusing the nodetool options in a different meaning is even more confusing for users with experience stock nodetool. For example:
> > > > > > ```
> > > > > >         -j <job_threads>, --job-threads <job_threads>
> > > > > >             Number of threads to run repair jobs. Usually this means number of
> > > > > >             CFs to repair concurrently. WARNING: increasing this puts more load
> > > > > >             on repairing nodes, so be careful. (default: 1, max: 4)
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > This is completely different from what you wanted the -j option to be below.
> > > > > > Instead of playing this game, IMO, we should deprecate nodetool all together and replace it with a more modern and clean tool.
> > > > > > > Like in #4848. Or, for `-j N`, making sure we are running at most `N` concurrent repairs on every node by strategically building the repair plan by splitting it into corresponding token ranges?
> > > > > > > How hard that would be?
> > > > > > 
> > > > > > 
> > > > > > Note nodetool is a single node tool. It is not a cluster wide management tool to schedule repair plans. Extending nodetool to do cluster wide management is way beyond the scope of this GH issue.
> > > > > 
> > > > > 
> > > > > We already have such a tool - called Scylla Manager. However, as you very well know, sometimes you still need an alternative.
> > > > > And it can be either a `nodetool` or a naked REST API. However we need a simple way to control the max repair concurrency (among other things). Why not to implement that? We know very well how to do that.
> > > > 
> > > > 
> > > > We can already do that with the existing restful api, i.e., issue a request with a single range. The value of allows specifying more ranges but allow only one range to work in parallel interally by introducing a new api parameter is limited, because 1) for advanced users, one range per request achieves the goal already (using the old repair script or write a new one) 2) normal user will not use this option at all.
> > > > A more advanced parameter like in #4848, which bypasses the system max_repair_ranges_in_parallel, allowing whatever parallelism specified by user. On one hand, it is more flexible for advanced users. On the other hand, it could cause problems when misused.
> > > 
> > > 
> > > @asias that's the whole point - this GH issue is not targeting "advanced users". It targets "simple people", like myself, who want to issue a single repair command and not bother who scylla does it's magic. And the same with restricting the parallelism: use a trivial "human readable" parameter to tell how many repairs per shard are allowed and let scylla handle the rest.
> > > Why can't we do that? (Again) Seems to be simple enough.
> > 
> > 
> > Normal users would not care about the internal parallelism has to be 1. The automatic parallelism should just work out of the box.
> 
> Exactly. But if it was 1 by default I would probably not even open this issue since it would "just work". But it seems it's not 1 but rather `max_repair_ranges_in_parallel` and this is exactly why I started this discussion.
> 
> "just work" or a way to control it otherwise is exactly what I'm asking here for.

With all the recent fixes, I think the default will just work. As I mentioned, there is already an way to control it, i.e., sending one range to repair at a time. 

> 
> > If it is not, we should fix the issues at the scylla core, e.g., the deadlock between the view update and streaming reader semaphore. So this is actually targeting advanced user cases to use this to workaround something. And you are not considered as "simple people" from any perspective ;-)
> 
> I'm speaking for "simple people" here... ;)

**Comment by vladzcloudius** [comment]
> With all the recent fixes, I think the default will just work.

Which "recent fixes", @asias? 
Are you saying that if I run `nodetool repair` on a node with all those fixes it will result in *up to one* repair context running on every shard on any node in the cluster at any given moment?

Again, `nodetool` users are not going to restrict ranges because they don't know and don't want to know what "range" even is.

**Comment by mykaul** [comment]
I /think/ what Asias refers to are the 2022.2.11 + 2022.2.12 fixes.
After all, this has been the default way of running repairs for a long time.

**Comment by nyh** [comment]
Duplicate of issues:
* https://github.com/scylladb/scylladb/issues/210 - Support sequential repair
* https://github.com/scylladb/scylladb/issues/2084 - nodetool repair option
* https://github.com/scylladb/scylladb/issues/12132 - this said we already fixed the documentation!

Someone who works on this issue, please try to close some of these issues as duplicates, and try to understand why we're still confused about this after the documentation improvements of #12132.

**Comment by asias** [comment]
> > With all the recent fixes, I think the default will just work.
> 
> Which "recent fixes", @asias? Are you saying that if I run `nodetool repair` on a node with all those fixes it will result in _up to one_ repair context running on any shard on any node in the cluster at any given moment?

No. I am talking about the fixes prevents the deadlock which happens more easily when more instances are running.  

1b577e0  (repair: Release permit earlier when the repair_reader is done)
19c3ecab (view_update_generator: Increase the registration_queue_size)

And fixes which make view build more efficient:

1d8cb32a5d (table: Optimize creation of reader excluding staging for view building)
a57724e71 (Make off-strategy compaction wait for view building completion)
3496b06aee (compaction: Prevent ICS from picking staging SSTables of a run)
169a8a66f28 (compatible_ring_position_or_view: make it cheap to copy)

Users should not care about if only one instance of repair is running on a shard. 

> 
> Again, `nodetool` users are not going to restrict ranges because they don't know and don't want to know what "range" even is.

It is documented. One can use nodetool repair -st end_token  -et end_token.

https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html

If they really do not want to do that and want a simpler solution,  they should use scylla manager which does everything.

**Comment by vladzcloudius** [comment]
> > > With all the recent fixes, I think the default will just work.
> > 
> > 
> > Which "recent fixes", @asias? Are you saying that if I run `nodetool repair` on a node with all those fixes it will result in _up to one_ repair context running on any shard on any node in the cluster at any given moment?
> 
> No. I am talking about the fixes prevents the deadlock which happens more easily when more instances are running.

These fixes are irrelevant to this GH issue.
Please, let's not mix all together.

> 
> [1b577e0](https://github.com/scylladb/scylladb/commit/1b577e0414547e775ef06370d1ed44c73492d517) (repair: Release permit earlier when the repair_reader is done) [19c3eca](https://github.com/scylladb/scylladb/commit/19c3ecabac265cc503269da14bb8b939d585d330) (view_update_generator: Increase the registration_queue_size)
> 
> And fixes which make view build more efficient:
> 
> [1d8cb32](https://github.com/scylladb/scylladb/commit/1d8cb32a5d467d483ae3b84af70b94eccfc3f847) (table: Optimize creation of reader excluding staging for view building) [a57724e](https://github.com/scylladb/scylladb/commit/a57724e711849f472a57d968fe07520ba6fac06e) (Make off-strategy compaction wait for view building completion) [3496b06](https://github.com/scylladb/scylladb/commit/3496b06aeea068371cbacad64c866bae4a27acc3) (compaction: Prevent ICS from picking staging SSTables of a run) [169a8a6](https://github.com/scylladb/scylladb/commit/169a8a66f282cb420197f446134c61760f1a087b) (compatible_ring_position_or_view: make it cheap to copy)
> 
> Users should not care about if only one instance of repair is running on a shard.
> 
> > Again, `nodetool` users are not going to restrict ranges because they don't know and don't want to know what "range" even is.
> 
> It is documented. One can use nodetool repair -st end_token -et end_token.
> 
> https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html
> 
> If they really do not want to do that and want a simpler solution, they should use scylla manager which does everything.

This issue is about a `nodetool`/REST API, not Scylla Manager. Please, let's stay focused and remain productive.

Again, I suggest to implement the repair logic on a single REST API call level.
The same logic used by SM today when it calls a repair REST API multiple times: given a "parallelism" value N repair master will pick both maximum internal concurrency and token ranges to repair so that there will be not more than N concurrent repairs (both as a slave and as a master) on any shard on any node at any time in the context of this particular repair task.

How hard would it be? (AFAICT - not very hard).
And if it's not very hard - why not to do that? 

(The fact that SM is supposed to do a similar thing by calling REST API multiple times is irrelevant).

@eliransin @asias Can I get an answer to this question, please?

**Comment by asias** [comment]
> > > > With all the recent fixes, I think the default will just work.
> > > 
> > > 
> > > Which "recent fixes", @asias? Are you saying that if I run `nodetool repair` on a node with all those fixes it will result in _up to one_ repair context running on any shard on any node in the cluster at any given moment?
> > 
> > 
> > No. I am talking about the fixes prevents the deadlock which happens more easily when more instances are running.
> 
> These fixes are irrelevant to this GH issue. Please, let's not mix all together.

They are because you hit the deadlock issue then you created this issue. 

> 
> > [1b577e0](https://github.com/scylladb/scylladb/commit/1b577e0414547e775ef06370d1ed44c73492d517) (repair: Release permit earlier when the repair_reader is done) [19c3eca](https://github.com/scylladb/scylladb/commit/19c3ecabac265cc503269da14bb8b939d585d330) (view_update_generator: Increase the registration_queue_size)
> > And fixes which make view build more efficient:
> > [1d8cb32](https://github.com/scylladb/scylladb/commit/1d8cb32a5d467d483ae3b84af70b94eccfc3f847) (table: Optimize creation of reader excluding staging for view building) [a57724e](https://github.com/scylladb/scylladb/commit/a57724e711849f472a57d968fe07520ba6fac06e) (Make off-strategy compaction wait for view building completion) [3496b06](https://github.com/scylladb/scylladb/commit/3496b06aeea068371cbacad64c866bae4a27acc3) (compaction: Prevent ICS from picking staging SSTables of a run) [169a8a6](https://github.com/scylladb/scylladb/commit/169a8a66f282cb420197f446134c61760f1a087b) (compatible_ring_position_or_view: make it cheap to copy)
> > Users should not care about if only one instance of repair is running on a shard.
> > > Again, `nodetool` users are not going to restrict ranges because they don't know and don't want to know what "range" even is.
> > 
> > 
> > It is documented. One can use nodetool repair -st end_token -et end_token.
> > https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html
> > If they really do not want to do that and want a simpler solution, they should use scylla manager which does everything.
> 
> This issue is about a `nodetool`/REST API, not Scylla Manager. Please, let's stay focused and remain productive.
> 
> Again, I suggest to implement the repair logic on a single REST API call level. The same logic used by SM today when it calls a repair REST API multiple times: given a "parallelism" value N repair master will pick both maximum internal concurrency and token ranges to repair so that there will be not more than N concurrent repairs (both as a slave and as a master) on any shard on any node at any time in the context of this particular repair task.

> 
> How hard would it be? (AFAICT - not very hard). And if it's not very hard - why not to do that?
> 
> (The fact that SM is supposed to do a similar thing by calling REST API multiple times is irrelevant).
> 
> @eliransin @asias Can I get an answer to this question, please?

Are you requesting implementing repair scheduling logic in scylla core instead of scylla manger?  To coordinate repair masters and followers, we need to coordinate cluster wide, not just the node that receives the rest api. I do not think it is a good idea to do that in scylla core. Such management fits best in the management tools not in scylla core. Scylla core already provides the mechanism to achieve that. 

Also the whole "N concurrent repairs" is not that important if we get rid of the streaming reader semaphore. 

reader_concurrency_semaphore: count resources are redundant  #13547

Limiting "N concurrent repairs" is just a workaround to mitigate the reader trash problem. We should fix the root cause, not changing the API to make the workaround easier to implement.

**Comment by asias** [comment]
> > > > > > With all the recent fixes, I think the default will just work.
> > > > > 
> > > > > 
> > > > > Which "recent fixes", @asias? Are you saying that if I run `nodetool repair` on a node with all those fixes it will result in _up to one_ repair context running on any shard on any node in the cluster at any given moment?
> > > > 
> > > > 
> > > > No. I am talking about the fixes prevents the deadlock which happens more easily when more instances are running.
> > > 
> > > 
> > > These fixes are irrelevant to this GH issue. Please, let's not mix all together.
> > 
> > 
> > They are because you hit the deadlock issue then you created this issue.
> 
> @asias I've opened this GH issue not only to address the deadlock issue you've referenced. The ability to control the repair concurrency has a much wider scope than that.
> 
> > > > [1b577e0](https://github.com/scylladb/scylladb/commit/1b577e0414547e775ef06370d1ed44c73492d517) (repair: Release permit earlier when the repair_reader is done) [19c3eca](https://github.com/scylladb/scylladb/commit/19c3ecabac265cc503269da14bb8b939d585d330) (view_update_generator: Increase the registration_queue_size)
> > > > And fixes which make view build more efficient:
> > > > [1d8cb32](https://github.com/scylladb/scylladb/commit/1d8cb32a5d467d483ae3b84af70b94eccfc3f847) (table: Optimize creation of reader excluding staging for view building) [a57724e](https://github.com/scylladb/scylladb/commit/a57724e711849f472a57d968fe07520ba6fac06e) (Make off-strategy compaction wait for view building completion) [3496b06](https://github.com/scylladb/scylladb/commit/3496b06aeea068371cbacad64c866bae4a27acc3) (compaction: Prevent ICS from picking staging SSTables of a run) [169a8a6](https://github.com/scylladb/scylladb/commit/169a8a66f282cb420197f446134c61760f1a087b) (compatible_ring_position_or_view: make it cheap to copy)
> > > > Users should not care about if only one instance of repair is running on a shard.
> > > > > Again, `nodetool` users are not going to restrict ranges because they don't know and don't want to know what "range" even is.
> > > > 
> > > > 
> > > > It is documented. One can use nodetool repair -st end_token -et end_token.
> > > > https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html
> > > > If they really do not want to do that and want a simpler solution, they should use scylla manager which does everything.
> > > 
> > > 
> > > This issue is about a `nodetool`/REST API, not Scylla Manager. Please, let's stay focused and remain productive.
> > > Again, I suggest to implement the repair logic on a single REST API call level. The same logic used by SM today when it calls a repair REST API multiple times: given a "parallelism" value N repair master will pick both maximum internal concurrency and token ranges to repair so that there will be not more than N concurrent repairs (both as a slave and as a master) on any shard on any node at any time in the context of this particular repair task.
> > 
> > 
> > > How hard would it be? (AFAICT - not very hard). And if it's not very hard - why not to do that?
> > > (The fact that SM is supposed to do a similar thing by calling REST API multiple times is irrelevant).
> > > @eliransin @asias Can I get an answer to this question, please?
> > 
> > 
> > Are you requesting implementing repair scheduling logic in scylla core instead of scylla manger? To coordinate repair masters and followers, we need to coordinate cluster wide, not just the node that receives the rest api. I do not think it is a good idea to do that in scylla core. Such management fits best in the management tools not in scylla core. Scylla core already provides the mechanism to achieve that.
> 
> Yes, I'm asking to implement a wiser logic of single repair steps in the scylla core. And unlike what SM does I'm only asking to implement the logic local to a specific repair master: if scylla receives a single repair REST API call for a range X which needs to be split in a certain way in order to achieve parallelism restrictions defined - scylla should do that.
> 
> That's not the same what SM does and this doesn't require a cluster-wide coordination since the cluster-wide coordination is outside the scope of a single REST API call.
> 
> In particular any concurrency guarantees are provided only in a scope of the specific REST API call. If somebody calls multiple such calls at the same time (e.g. from different nodes) the total repair concurrency on a single shard may be larger than the one guaranteed by a single REST API call.
> 
> I hope it's clearer now what I actually intended.

This is what we already did in https://github.com/scylladb/scylladb/pull/4848. We can revive this PR in case it is useful in the field.  I do not want the user specified parallelism to exceed the system max_repair_ranges_in_parallel though. 

> > Also the whole "N concurrent repairs" is not that important if we get rid of the streaming reader semaphore.
> > reader_concurrency_semaphore: count resources are redundant #13547
> > Limiting "N concurrent repairs" is just a workaround to mitigate the reader trash problem. We should fix the root cause, not changing the API to make the workaround easier to implement.
> 
> Again, Asias, your assumption that this issue's scope is limited to a specific problem (workaround) is baseless.

**Comment by vladzcloudius** [comment]
> This is what we already did in https://github.com/scylladb/scylladb/pull/4848. We can revive this PR in case it is useful in the field. I do not want the user specified parallelism to exceed the system max_repair_ranges_in_parallel though.

That's exactly what I was looking for, @asias! Thanks you!
I re-opened the PR and posted a clarifying comment.
